lib/std/os.zig

```diff
@@ -4152,6 +4156,71 @@ pub fn getFdPath(fd: fd_t, out_buffer: *[MAX_PATH_BYTES]u8) RealPathError![]u8 {
             const len = mem.indexOfScalar(u8, out_buffer[0..], @as(u8, 0)) orelse MAX_PATH_BYTES;
             return out_buffer[0..len];
         },
+        .freebsd => {
+            const pid = @intCast(c_int, system.getpid());
+
+            var mib = [_]c_int{ 0, 0, 0, pid };
+            var miblen = mib.len;
+            // Pending https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=198570
+            // or similar, there's not a good way to get this in FreeBSD.
+            // Instead, list all open files for the process and find the FD we're looking for.
+            switch (errno(system.sysctlnametomib("kern.proc.filedesc", &mib[0], &miblen))) {
+                0 => {},
+                EFAULT => unreachable, // stack is always a valid address
+                EINVAL => unreachable, // known good name
+                ENOMEM => unreachable, // known size, on the stack
+                EPERM => unreachable, // not setting new value
+                ENOENT => return error.NotSupported,
+                else => |err| return unexpectedErrno(err),
+            }
+
+            // If the buffer is too small, sysctl silently gives partial result,
+            // which may not contain our target FD.
+            // Get the approximate required size by using null for oldp/buf.
+            // Use that to change the error produced if we can't find the FD.
+            // TODO: Do this better with an allocator, alloca or stupidly large array?
+            var buflen: usize = 0;
+            sysctl(&mib, null, &buflen, null, 0) catch |err| switch (err) {
+                error.PermissionDenied => unreachable, // not setting value
+                error.NameTooLong => unreachable, // known good mib from sysctlnametomib
+                error.UnknownName => unreachable, // known good mib from sysctlnametomib
+                error.Unexpected => return error.Unexpected, // EINVAL/EISDIR/ENOTDIR should be unreachable
+                error.SystemResources => return error.SystemResources,
+            };
+
+            // 20 is completely arbitrary, should work to around 40 open files?
+            var buf: [MAX_PATH_BYTES * 20]u8 = undefined;
```

Comment:
@andrewrk might correct me here, but I think you can specify the alignment right here as

```zig
var buf: [MAX_PATH_BYTES * 20]u8 align(@alignOf(system.kinfo_file)) = undefined;
```

Then you should be able to `@ptrCast` directly later on without `@alignCast`s like so

```zig
const kf = @ptrCast(*system.kinfo_file, &buf[offset]);
```

Reply:
This is extremely confusing, what does `MAX_PATH_BYTES` have anything to do with an array of `system.kinfo_file`?
Since the `kinfo_file` has a worst-case size of `@sizeOf(system.kinfo_file)` you may want to allocate `K` times that.

Reply:
On second thought though @LemonBoy, I'm guessing here, but perhaps `kinfo_file` is a VLA, and hence, we should allocate `K` times `kinfo_file + MAX_PATH_BYTES`?

Reply:
[kinfo_getfile](https://github.com/freebsd/freebsd/blob/master/lib/libutil/kinfo_getfile.c) suggests `kinfo_file` size is the upper boundary, `kf_structsize` should be less or equal to its size.

Reply:
Total derp on my part, you're completely correct @LemonBoy, thanks!